src/wrappers/pubsub.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * @fileoverview Handlers for Pubsub instrumentation
+ */
+
+const shimmer = require('shimmer');
+const {
+    tracer,
+    tryRequire,
+    eventInterface,
+} = require('epsagon');
+const traceContext = require('../trace_context.js');
+
+const subscriber = tryRequire('@google-cloud/pubsub/build/src/subscriber');
+
+/**
+ * Handle subscriber event emitter of eventName='message'
+ * @param {Message} message received message.
+ * @param {*} originalHandler listener callback function.
+ * @param {*} requestFunctionThis request arguments.
+ */
+function pubSubSubscriberMiddleware(message, originalHandler, requestFunctionThis) {
+    let originalHandlerErr;
+    try {
+        // Initialize tracer and evnets.
+        tracer.restart();
+        const functionName = originalHandler.name || 'messageHandler';
+        const { slsEvent: pubSubEvent, startTime: pubSubStartTime } =
+        eventInterface.initializeEvent(
+            'pubsub',
+            requestFunctionThis.projectId,
+            'messagePullingListener',
+            'trigger'
+        );
+        const { slsEvent: nodeEvent, startTime: nodeStartTime } = eventInterface.initializeEvent(
+            'node_function', functionName, 'messageReceived', 'runner'
+        );
+        tracer.addEvent(pubSubEvent);
+
+        // Getting message data.
+        let callbackResponse = { messageId: message.id };
+        const messageData = (message.data && JSON.parse(`${message.data}`));
+        if (messageData && typeof messageData === 'object') {
+            callbackResponse = Object.assign(callbackResponse, messageData);
+        }
+        const { label, setError } = tracer;
+        const returnMessage = message;
+        returnMessage.epsagon = {
+            label,
+            setError,
+        };
+
+        // Finalize pubsub event.
+        eventInterface.finalizeEvent(pubSubEvent, pubSubStartTime, null, callbackResponse);
```

Comment:
finalize it before starting to deal with the runner event (you want the duration to be 0 anyway)